src/networking.c

```diff
@@ -2179,44 +2519,89 @@ int _writeToClient(client *c) {
     }
 
     c->nwritten = tot_written;
+    if (tot_written > 0) {
+        c->io_last_written_buf = c->buf;
+        c->io_last_written_bufpos = (tot_written == bytes_to_write ? bufpos : 0);
+        c->io_last_written_data_len = c->io_last_written_data_len + tot_written;
+    }
     return tot_written > 0 ? C_OK : C_ERR;
 }
 
-static void _postWriteToClient(client *c) {
-    if (c->nwritten <= 0) return;
+void resetLastWrittenBuf(client *c) {
+    c->io_last_written_buf = NULL;
+    c->io_last_written_bufpos = 0;
+    c->io_last_written_data_len = 0;
+}
+
+static void releaseBufOffloads(char *buf, size_t bufpos) {
+    char *ptr = buf;
+    while (ptr < buf + bufpos) {
+        payloadHeader *header = (payloadHeader *)ptr;
+        ptr += sizeof(payloadHeader);
+
+        if (header->type == CLIENT_REPLY_PAYLOAD_BULK_OFFLOAD) {
+            clusterSlotStatsAddNetworkBytesOutForSlot(header->slot, header->actual_len);
+
+            robj** obj_ptr = (robj**)ptr;
```

Comment:
Code like this would benefit a lot from some helper methods. Instead of just constantly moving and recasting values. Something like,

```
robj *getValkeyObjectFromHeader(payloadHeader *header) {
     char *ptr = (char *ptr) header;
     ptr += sizeof(payloadHeader);
     return (robj**)ptr;
}
```